server/core/terraform/terraform_client.go

```diff
@@ -262,6 +279,83 @@ func (c *DefaultClient) TerraformBinDir() string {
 	return c.binDir
 }
 
+// ListAvailableVersions returns all available version of Terraform. If downloads are not allowed, this will return an empty list.
+func (c *DefaultClient) ListAvailableVersions(log logging.SimpleLogging) ([]string, error) {
+	url := fmt.Sprintf("%s/terraform", c.downloadBaseURL)
+
+	if !c.downloadsAllowed {
+		log.Debug("Terraform downloads disabled. Won't list Terraform versions available at %s", url)
+		return []string{}, nil
+	}
+
+	log.Debug("Listing Terraform versions available at: %s", url)
+	versions, err := lib.GetTFList(url, true)
+	return versions, err
+}
+
+// DetectVersion Extracts required_version from Terraform configuration in the specified project directory. Returns nil if unable to determine the version.
+// This will also try to intelligently evaluate non-exact matches by listing the available versions of Terraform and picking the best match.
+func (c *DefaultClient) DetectVersion(projectDirectory string, log logging.SimpleLogging) *version.Version {
+	module, diags := tfconfig.LoadModule(projectDirectory)
+	if diags.HasErrors() {
+		log.Err("trying to detect required version: %s", diags.Error())
+	}
+
+	if len(module.RequiredCore) != 1 {
+		log.Info("cannot determine which version to use from terraform configuration, detected %d possibilities.", len(module.RequiredCore))
+		return nil
+	}
+	requiredVersionSetting := module.RequiredCore[0]
+	log.Debug("found required_version setting of %q", requiredVersionSetting)
+
+	tfVersions, err := c.ListAvailableVersions(log)
+	if err != nil {
+		log.Err("Unable to list Terraform versions, may fall back to default: %s", err)
+	}
+
+	if len(tfVersions) == 0 {
+		// Fall back to an exact required version string
+		// We allow `= x.y.z`, `=x.y.z` or `x.y.z` where `x`, `y` and `z` are integers.
+		re := regexp.MustCompile(`^=?\s*([0-9.]+)\s*$`)
+		matched := re.FindStringSubmatch(requiredVersionSetting)
+		if len(matched) == 0 {
+			log.Debug("Did not specify exact version in terraform configuration, found %q", requiredVersionSetting)
+			return nil
+		}
+		tfVersions = []string{matched[1]}
+	}
+
+	constraint, _ := semver.NewConstraint(requiredVersionSetting)
+	versions := make([]*semver.Version, len(tfVersions))
+
+	for i, tfvals := range tfVersions {
+		newVersion, err := semver.NewVersion(tfvals) //NewVersion parses a given version and returns an instance of Version or an error if unable to parse the version.
```

Comment:
nit comment on new line for readability
```suggestion
		// NewVersion parses a given version and returns an instance of Version or an error if unable to parse the version.
		newVersion, err := semver.NewVersion(tfvals)
```

Reply:
I just removed this since it's a copy of the documentation for the called function.